### ISSUE TITLE
invalid use of incomplete type 'class DICOMParser'

### ISSUE THREAD

**Issue by oroppas** [vmtk/vmtk#198]
Hi,

I've got the error while I'm building VMTK on msys2 using mingw-w64 64-bit compiler

$ gcc --version
gcc.exe (Rev1, Built by MSYS2 project) 7.2.0

$ which gcc
/mingw64/bin/gcc

```
[1/23] Building CXX object vtkVmtk/IO/CMakeFiles/vtkvmtkIO.dir/vtkvmtkDICOMImageReader.cxx.obj
FAILED: vtkVmtk/IO/CMakeFiles/vtkvmtkIO.dir/vtkvmtkDICOMImageReader.cxx.obj
C:\msys64\mingw64\bin\c++.exe  -DITK_IO_FACTORY_REGISTER_MANAGER -DvtkvmtkIO_EXPORTS -isystem C:/msys64/mingw64/include/vtk-8.1 -IITKIOFactoryRegistration -IC:/msys64/mingw64/include/gdcm-2.8 -IC:/msys64/mingw64/include -IC:/msys64/mingw64/include/ITK-4.13 -IC:/msys64/home/cd4mcq/packages/vmtk/src/vmtk/vtkVmtk -IC:/msys64/home/cd4mcq/packages/vmtk/src/vmtk/vtkVmtk/Common -IC:/msys64/home/cd4mcq/packages/vmtk/src/vmtk/vtkVmtk/ComputationalGeometry -IC:/msys64/home/cd4mcq/packages/vmtk/src/vmtk/vtkVmtk/DifferentialGeometry -IC:/msys64/home/cd4mcq/packages/vmtk/src/vmtk/vtkVmtk/IO -IC:/msys64/home/cd4mcq/packages/vmtk/src/vmtk/vtkVmtk/Misc -IC:/msys64/home/cd4mcq/packages/vmtk/src/vmtk/vtkVmtk/Rendering -IC:/msys64/home/cd4mcq/packages/vmtk/src/vmtk/vtkVmtk/Segmentation -IC:/msys64/home/cd4mcq/packages/vmtk/src/vmtk/vtkVmtk/Contrib -IC:/msys64/home/cd4mcq/packages/vmtk/src/vmtk/vtkVmtk/Utilities/vtkvmtkITK -IvtkVmtk -march=x86-64 -mtune=generic -O2 -pipe  -msse2 -mthreads  -msse2 -mthreads  -mwin32 -mthreads -O3 -DNDEBUG   -std=gnu++11 -MD -MT vtkVmtk/IO/CMakeFiles/vtkvmtkIO.dir/vtkvmtkDICOMImageReader.cxx.obj -MF vtkVmtk\IO\CMakeFiles\vtkvmtkIO.dir\vtkvmtkDICOMImageReader.cxx.obj.d -o vtkVmtk/IO/CMakeFiles/vtkvmtkIO.dir/vtkvmtkDICOMImageReader.cxx.obj -c C:/msys64/home/cd4mcq/packages/vmtk/src/vmtk/vtkVmtk/IO/vtkvmtkDICOMImageReader.cxx
C:/msys64/home/cd4mcq/packages/vmtk/src/vmtk/vtkVmtk/IO/vtkvmtkDICOMImageReader.cxx: In member function 'void vtkvmtkDICOMImageReader::ComputeOutputVoxelSpacing()':
C:/msys64/home/cd4mcq/packages/vmtk/src/vmtk/vtkVmtk/IO/vtkvmtkDICOMImageReader.cxx:90:17: error: invalid use of incomplete type 'class DICOMParser'
     this->Parser->ClearAllDICOMTagCallbacks();
                 ^~
In file included from C:/msys64/home/cd4mcq/packages/vmtk/src/vmtk/vtkVmtk/IO/vtkvmtkDICOMImageReader.h:29:0,
                 from C:/msys64/home/cd4mcq/packages/vmtk/src/vmtk/vtkVmtk/IO/vtkvmtkDICOMImageReader.cxx:22:
C:/msys64/mingw64/include/vtk-8.1/vtkDICOMImageReader.h:44:7: note: forward declaration of 'class DICOMParser'
 class DICOMParser;
       ^~~~~~~~~~~
...
```

**Comment by rlizzo** [comment]
Hi @oroppas. I haven't been able to replicate this issue. what version of VTK were you compiling against?

**Comment by oroppas** [comment]
Hi @rlizzo,

I'm compiling against VTK 8.1.0

My temporary fix is
```
//class DICOMParser;
//class DICOMAppHelper;
#include "DICOMParser.h"
#include "DICOMAppHelper.h"
```
this looks like a issue on VTK side.

**Comment by rlizzo** [comment]
Thanks for the update @oroppas! I just tried it, and I agree, it seems to be an issue with VTK 8.1 as it works with everything up to VTK 8.0.1. I'll put in an update with your fix tonight. Thanks for reporting!

**Comment by oroppas** [comment]
@rlizzo. Pardon me I wan't clear about which file I was talking about.
The file I applied the fix was ```vtk-8.1/vtkDICOMImageReader.h```.

Do you think it's a bug in VTK?

**Comment by rlizzo** [comment]
@oroppas, thanks for the clarification. 

@lantiga any ideas? I can recreate the error, but aside from the vtk patch suggested above, haven't been able to find any workaround for this.

**Comment by lantiga** [comment]
Hey @rlizzo! Thanks @oroppas for the feedback.

So, the issue is that the class if forward-declared (no explicit import) but then one of its members are dereferenced further down. The fact that it's manifesting here might be due to a stricter behavior by the compiler, I'm not sure.

Anyway, try adding
```
#include "DICOMParser.h"
#include "DICOMAppHelper.h"
```
right before `#include "vtkDICOMImageReader.h"` in `vmtk/vtkVmtk/IO/vtkvmtkDICOMImageReader.h`. This might do the trick without patching VTK.

**Comment by rlizzo** [comment]
Thanks @lantiga. Unfortunately that doesn't work out. 

We get the following error instead:

```cpp
/Users/rick/projects/vmtk/vmtk/vtkVmtk/IO/vtkvmtkDICOMImageReader.cxx:90:17: error: member access into incomplete type
      'DICOMParser'
    this->Parser->ClearAllDICOMTagCallbacks();
                ^
/Users/rick/anaconda3/envs/buildvmtknew/include/vtk-8.1/vtkDICOMImageReader.h:44:7: note: forward declaration of
      'DICOMParser'
class DICOMParser;
```

**Comment by lantiga** [comment]
Ouch, this is weird. For some reason, the forward declarations in `vtkDICOMImageReader` are overriding the full definitions that are included when we `#include "DICOMParser.h"`.

The original code was correct: we first included `#include "vtkDICOMImageReader.h"` in the header file, which contained the forward declaration, and later we included `#include "DICOMParser.h"` in `vtkDICOMImageReader.cxx`, which contains the full definition. For some reason this isn't working...
